Categories

abhikanojia opened the pull request.

Commit:
Started Categories and subcategories implementation

Commit:
Minor changes

abhishekkanojia committed:
Added self relation to category model.
Added category dropdown to products form, changed controller accordingly.
added new migration for adding category_id to products.

Commit:
Removed ul css from scaffold scss.
added action categories to products controller and added route for same.
linked category and product models.
added category option to products form
renamed count to products count migration 1
changed default from nil to 0 migration 2

Commit:
Added translation for category_prompt on products page
changed order of includes in product controller

harsh-ag commented on depot/app/models/product.rb:
```diff
@@ -9,6 +9,7 @@ class Product < ApplicationRecord
   has_many :line_items, dependent: :restrict_with_error
   has_many :orders, through: :line_items
   has_many :carts, through: :line_items
+  belongs_to :category, counter_cache: true
```
is this working fine.?

harsh-ag commented on depot/app/models/category.rb:
```diff
@@ -0,0 +1,43 @@
+  before_save :ensure_not_referring_to_sub_category
+
+  def sub_category_exists?
+    !sub_categories.empty?
```
better way

harsh-ag commented on depot/app/models/category.rb:
```diff
@@ -0,0 +1,43 @@
+
+  private
+
+    def increment_count_in_parent_category
```
what is the need of this.? Plz go through your logic once

harsh-ag commented on depot/db/migrate/20171219124729_create_categories.rb:
```diff
@@ -0,0 +1,11 @@
+    create_table :categories do |t|
+      t.string :name, null: false
+      t.integer :count
+      t.references :parent_category
```
shouldn't it be parent_id?

harsh-ag commented on depot/app/models/category.rb:
```diff
@@ -0,0 +1,43 @@
+  # validations
+  validates :name, presence: true
+  validates :name, uniqueness: { scope: :parent_category_id }, allow_blank: true
+  validates_uniqueness_of :name, if: :name?
```
better way

Commit:
Refactored code.
Added code to increment parent category count
new migration to change name of parent_category_id to parent_id

Commit:
Need to change in category callbacks for decrement.

Commit:
Minor fixes, to category and product models.

harsh-ag commented on depot/app/controllers/products_controller.rb:
```diff
@@ -80,6 +80,10 @@ def who_bought
     end
   end
 
+  def categories
```
separate controller for categories

harsh-ag commented on depot/app/models/category.rb:
```diff
@@ -0,0 +1,38 @@
+  private
+
+    def ensure_no_products_exists
+      if !products.empty?
```
can use unles or products.present?

harsh-ag commented on depot/app/models/category.rb:
```diff
@@ -0,0 +1,38 @@
+      end
+    end
+
+    def ensure_not_referring_to_sub_category
```
it should be a callback or validation..??

harsh-ag commented on depot/app/models/product.rb:
```diff
@@ -9,10 +9,15 @@ class Product < ApplicationRecord
+  # callbacks
   before_validation :set_default_title, unless: :title?
   before_validation :set_default_discount_price, unless: :discount_price?
+  after_create_commit :increment_count_in_parent_category, if: :category_id?
```
do as discussed

Commit:
Created new categories controller.
Refactored code.

Commit:
Changed parent_category_id to parent_id

Commit:
Minor fixes
